pkg/controller/elasticsearch/shutdown/node.go

```diff
@@ -142,13 +151,17 @@ func logStatus(logger logr.Logger, podName string, shutdown esclient.NodeShutdow
 
 // Clear deletes shutdown requests matching the type of the NodeShutdown field typ and the given optional status.
 // Depending on the progress of the shutdown in question this means either a cancellation of the shutdown or a clean-up
-// after shutdown completion.
+// after shutdown completion. Restart shutdowns will only be deleted once the corresponding node is back in the cluster.
 func (ns *NodeShutdown) Clear(ctx context.Context, status *esclient.ShutdownStatus) error {
 	if err := ns.initOnce(ctx); err != nil {
 		return err
 	}
 	for _, s := range ns.shutdowns {
 		if s.Is(ns.typ) && (status == nil || s.Status == *status) {
+			if ns.typ == esclient.Restart && !ns.nodeInCluster(s.NodeID) {
```

Comment:
Not related to your PR but I never realized that `Clear()` may never be called until all nodes are restarted: https://github.com/elastic/cloud-on-k8s/blob/757a07eb557b992a4d1645ce86eb8d5650907f5e/pkg/controller/elasticsearch/driver/upgrade.go#L111-L112
I managed to reproduce a situation where the 3 nodes of a cluster are restarted, but `Clear()` is never called during the upgrade, leading to a situation where restarted nodes are piling up in the cluster state:

```json
{
	"nodes": [{
		"node_id": "B_CBcBIzTlGnKrereheimw",
		"type": "RESTART",
		"reason": "610898",
		"allocation_delay": "5m",
		"shutdown_startedmillis": 1650956474657,
		"status": "COMPLETE",
		"shard_migration": {
			"status": "COMPLETE",
			"shard_migrations_remaining": 0,
			"explanation": "no shard relocation is necessary for a node restart"
		},
		"persistent_tasks": {
			"status": "COMPLETE"
		},
		"plugins": {
			"status": "COMPLETE"
		}
	}, {
		"node_id": "Htx8TgKwS5CE6Mt69SVcug",
		"type": "RESTART",
		"reason": "612171",
		"allocation_delay": "5m",
		"shutdown_startedmillis": 1650956680896,
		"status": "COMPLETE",
		"shard_migration": {
			"status": "COMPLETE",
			"shard_migrations_remaining": 0,
			"explanation": "no shard relocation is necessary for a node restart"
		},
		"persistent_tasks": {
			"status": "COMPLETE"
		},
		"plugins": {
			"status": "COMPLETE"
		}
	}, {
		"node_id": "w3R9ACQ8RseE9cEQDJf8bQ",
		"type": "RESTART",
		"reason": "611345",
		"allocation_delay": "5m",
		"shutdown_startedmillis": 1650956571498,
		"status": "COMPLETE",
		"shard_migration": {
			"status": "COMPLETE",
			"shard_migrations_remaining": 0,
			"explanation": "no shard relocation is necessary for a node restart"
		},
		"persistent_tasks": {
			"status": "COMPLETE"
		},
		"plugins": {
			"status": "COMPLETE"
		}
	}]
}
```

I'm wondering if we should consider it as a bug?

Reply:
I think potentially yes. It could become a problem if another restart is necessary before the old records have been cleared. I think keeping these records in the cluster state is probably less of a problem in terms of size. I feel like we should probably tackle this in a separate issue though and it probably has a little bit lower priority than this one wdyt?

Reply:
> I feel like we should probably tackle this in a separate issue though and it probably has a little bit lower priority than this one wdyt?

Yes, absolutely 👍